src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -2368,15 +2369,15 @@ private static void FormatPercent(ref ValueStringBuilder sb, ref NumberBuffer nu
             }
         }
 
-        internal static unsafe void RoundNumber(ref NumberBuffer number, int pos)
+        internal static unsafe void RoundNumber(ref NumberBuffer number, int pos, bool isCorrectlyRounded)
         {
             byte* dig = number.GetDigitsPointer();
 
             int i = 0;
-            while (i < pos && dig[i] != 0)
+            while (i < pos && dig[i] != '\0')
                 i++;
 
-            if (i == pos && dig[i] >= '5')
+            if ((i == pos) && ShouldRoundUp(dig, i, number.Kind, isCorrectlyRounded))
```

Comment:
Post 3.0, we could fixup the handling for custom numeric-format specifiers to correctly determine how many digits are being requested and if the user is requesting significant digits or a number of digits that appear after the decimal point.

If we have that information, then `Grisu3`/`Dragon4` will correctly round the generated digits themselves and we can simplify this back to `if ((i == pos) && (dig[i] >= '5') && (number.Kind != NumberBufferKind.FloatingPoint))`